packages/firestore/src/api/database.ts

```diff
@@ -1942,77 +1361,59 @@ export class CollectionReference<T = PublicDocumentData>
   extends Query<T>
   implements PublicCollectionReference<T> {
   constructor(
-    readonly _path: ResourcePath,
-    firestore: Firestore,
-    _converter: UntypedFirestoreDataConverter<T> | null
+    readonly firestore: Firestore,
+    readonly _delegate: ExpCollectionReference<T>
   ) {
-    super(newQueryForPath(_path), firestore, _converter);
-    if (_path.length % 2 !== 1) {
-      throw new FirestoreError(
-        Code.INVALID_ARGUMENT,
-        'Invalid collection reference. Collection ' +
-          'references must have an odd number of segments, but ' +
-          `${_path.canonicalString()} has ${_path.length}`
-      );
-    }
+    super(firestore, _delegate);
   }
 
   get id(): string {
-    return this._query.path.lastSegment();
+    return this._delegate.id;
   }
 
-  get parent(): PublicDocumentReference<PublicDocumentData> | null {
-    const parentPath = this._query.path.popLast();
-    if (parentPath.isEmpty()) {
-      return null;
-    } else {
-      return DocumentReference.forPath(
-        parentPath,
-        this.firestore,
-        /* converter= */ null
-      );
-    }
+  get path(): string {
+    return this._delegate.path;
   }
 
-  get path(): string {
-    return this._query.path.canonicalString();
+  get parent(): DocumentReference<PublicDocumentData> | null {
+    const docRef = this._delegate.parent;
+    return docRef ? new DocumentReference(this.firestore, docRef) : null;
   }
 
-  doc(pathString?: string): PublicDocumentReference<T> {
-    // We allow omission of 'pathString' but explicitly prohibit passing in both
-    // 'undefined' and 'null'.
-    if (arguments.length === 0) {
-      pathString = AutoId.newId();
+  doc(documentPath?: string): DocumentReference<T> {
+    try {
+      if (documentPath !== undefined) {
+        return new DocumentReference(
```

Comment:
Why doesn't this case work if `documentPath` is undefined? I thought `doc` can't tell if you called it as `doc(this._delegate, undefined)` vs `doc(this._delegate)`.

Reply:
I believe we owe this to the genius of @mikelehen.

I added this comment:

```
        // Call `doc` without `documentPath` if `documentPath` is `undefined`
        // as `doc` validates the number of arguments to prevent users from
        // accidentally passing `undefined`.
```

Reply:
😱